lib/krane/deprecated_deploy_task.rb

```diff
@@ -447,6 +446,33 @@ def deploy_resources(resources, prune: false, verify:, record_summary: true)
       end
     end
 
+    def replace_or_create_resource(resource, force: false)
+      args = if force
+        ["replace", "--force", "--cascade", "-f", resource.file_path]
+      else
+        ["replace", "-f", resource.file_path]
+      end
+
+      _, _, status = kubectl.run(*args, log_failure: false)
+      return [nil, status] if status.success?
+
+      # it doesn't exist so we can't replace it, we try to create it
+      create_resource(resource)
+    end
+
+    def create_resource(resource)
+      out, err, status = kubectl.run("create", "-f", resource.file_path, log_failure: false, output: 'json',
+        output_is_sensitive: resource.sensitive_template_content?)
+
+      # For resources that rely on a generateName attribute, we get the `name` from the result of the call to `create`
+      # We must explicitly set this name value so that the `apply` step for pruning can run successfully
+      if status.success? && resource.deploy_method != :apply
```

Comment:
If we're here, don't we already know `resource.deploy_method != :apply`? This will make us set the generated name in a lot of cases we don't need to. Why not directly check whether the resource used generateName? Two options come to mind:
A - expose a `uses_generate_name?` method on the resource that looks at `@definition`
B - parse the JSON here and dig for `generateName`. Then also dig out the real name and pass that in instead of the full JSON.

B probably looks nicer and reduces the API surface of resources... but then again it would make us parse JSON when we don't need to. So I'm inclined to go with A.